ddht/v5_1/alexandria/client.py

```diff
@@ -496,6 +531,39 @@ async def advertise(
             ]
         )
 
+    async def locate(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        *,
+        content_key: ContentKey,
+        request_id: Optional[bytes] = None,
+    ) -> Tuple[InboundMessage[LocationsMessage], ...]:
+        request = LocateMessage(LocatePayload(content_key))
+
+        subscription: trio.abc.ReceiveChannel[InboundMessage[LocationsMessage]]
+        # unclear why `subscribe_request` isn't properly carrying the type information
+        async with self.subscribe_request(  # type: ignore
+            node_id, endpoint, request, LocationsMessage, request_id=request_id,
+        ) as subscription:
+            head_response = await subscription.receive()
+            total = head_response.message.payload.total
+            responses: Tuple[InboundMessage[LocationsMessage], ...]
+            if total == 1:
+                responses = (head_response,)
+            elif total > 1:
+                tail_responses: List[InboundMessage[LocationsMessage]] = []
+                for _ in range(total - 1):
+                    tail_responses.append(await subscription.receive())
```

Comment:
What happens if for whatever reason one of the messages gets lost in transmission?

Reply:
Then the whole request errors out.  It's not ideal because *probably* the incomplete set of other responses contain useful information...  Which suggests that we probably need to do roughly the same pattern here as is done with `NetworkAPI.recursive_find_nodes` and to stream the response ENR records asynchronously.

```python
async with network.find_nodes(...) as enrs_aiter:
    async for enr in enrs_aiter:
        ...
```

But this is subject to all of the necessary exception handling so it turns into:

```python
try:
    async with network.find_nodes(...) as enrs_aiter:
        async for enr in enrs_aiter:
            ...
except (trio.TooSlowError, ValidationError):
    ...
```

I think this would yield a performance improvement to RFN as well since it would allow workers to consume the ENRS as soon as they are received rather than having to wait until all of them have been received.